cocos/scene-graph/node.ts

```diff
@@ -2565,24 +2565,55 @@ export class Node extends CCObject implements ISchedulable, CustomSerializable {
 
             Vec3.copy(worldScale, v3_a);
         }
+
+        let rotationFlag = TransformBit.NONE;
         if (parent) {
-            v3_a.x = worldScale.x / Vec3.set(v3_b, this._mat.m00, this._mat.m01, this._mat.m02).length();
-            v3_a.y = worldScale.y / Vec3.set(v3_b, this._mat.m04, this._mat.m05, this._mat.m06).length();
-            v3_a.z = worldScale.z / Vec3.set(v3_b, this._mat.m08, this._mat.m09, this._mat.m10).length();
+            const xScale = Vec3.set(v3_b, this._mat.m00, this._mat.m01, this._mat.m02).length();
+            const yScale = Vec3.set(v3_b, this._mat.m04, this._mat.m05, this._mat.m06).length();
+            const zScale = Vec3.set(v3_b, this._mat.m08, this._mat.m09, this._mat.m10).length();
+            if (xScale === 0) {
+                v3_a.x = worldScale.x;
+                this._mat.m00 = 1;
+                rotationFlag = TransformBit.ROTATION;
```

Comment:
Why scale effects rotation?

Reply:
![WeChat4a7135b19bb17b607d38e1354624c7d2](https://github.com/user-attachments/assets/5dd32fe7-30b6-4698-9965-1a9d68dce3c4)
If Sx, Sy, or Sz is zero, the original Rij (The elements of the rotation matrix) will be lost since every number multipy zero is zero.

And in our engine code, Node.prototype.updateWorldTransform:
```ts
if (dirtyBits & TransformBit.POSITION) {
    Vec3.transformMat4(child._pos, child._lpos, cur._mat);
    child._mat.m12 = child._pos.x;
    child._mat.m13 = child._pos.y;
    child._mat.m14 = child._pos.z;
}
if (dirtyBits & TransformBit.RS) {
    // If child._lscale `(0, 0, 0)`, child._mat will be [0, 0, 0, 0,   0, 0, 0, 0,   0, 0, 0, 0,    0, 0, 0, 1], the matrix wil be degenerated since scale is zero since rotation elements in matrix will be lost.
    Mat4.fromRTS(child._mat, child._lrot, child._lpos, child._lscale);
    Mat4.multiply(child._mat, cur._mat, child._mat);

    // If the TransformBit.ROTATION is not set, child._rot (world rotation) will not be updated.
    const rotTmp = dirtyBits & TransformBit.ROTATION ? child._rot : null;
    // In toRTS, since if one of Sx, Sy, Sz is zero, the rotTmp will be reset to `Unit Quaternion`.
    Mat4.toRTS(child._mat, rotTmp, null, child._scale);
}
```

```ts
public static toSRT<InType extends IMat4Like, VecLike extends IVec3Like> (m: InType, q: Quat | null, v: VecLike | null, s: VecLike | null): void {
......

        // Scale
        const sx = Vec3.set(v3_1, m.m00, m.m01, m.m02).length();
        const sy = Vec3.set(v3_1, m.m04, m.m05, m.m06).length();
        const sz = Vec3.set(v3_1, m.m08, m.m09, m.m10).length();
        if (s) {
            s.x = sx;
            s.y = sy;
            s.z = sz;
        }

        // Scale too close to zero, can't decompose rotation.
        if (sx === 0 || sy === 0 || sz === 0) {
            if (q) {
                Quat.identity(q);
            }
            return;
        }
```

Reply:
Think of this invocations:
1. setRotationFromEuler(10, 10, 10);
2. setWorldScale(0, 0, 0); // Rotation elements are lost
3. setRotationFromEuler(20, 20, 20); // Change the local rotation quaternion
4. setWorldScale(1, 1, 1); // Set the world scale to correct values, if not setting the TransformBit.ROTATION flag, the following updateWorldTransform will not update _rot (world rotation) since `const rotTmp = dirtyBits & TransformBit.ROTATION ? child._rot : null;`
5. updateWorldTransform();

Reply:
> setWorldScale(0, 0, 0); // Rotation elements are lost

With this PR, the rotation will not be lost, right?

Reply:
After merging this PR, while setWorldScale(1, 1, 1) to non-zero values, the world rotation will be re-caculated and set to the correct value since we mark the TransformBit.ROTATION in `setWorldScale` which means that we need to update the world rotation in updateWorldTransform.